cores/esp32/esp32-hal-ledc.c

```diff
@@ -119,12 +134,12 @@ bool ledcAttachChannel(uint8_t pin, uint32_t freq, uint8_t resolution, uint8_t c
 
 bool ledcAttach(uint8_t pin, uint32_t freq, uint8_t resolution)
 {
-    uint8_t free_channel = ~ledc_handle.used_channels & (ledc_handle.used_channels+1);
-    if (free_channel == 0 || resolution > LEDC_MAX_BIT_WIDTH){
-        log_e("No more LEDC channels available! (maximum %u) or bit width too big (maximum %u)", LEDC_CHANNELS, LEDC_MAX_BIT_WIDTH);
+    int free_channel = ~ledc_handle.used_channels & (ledc_handle.used_channels+1);
```

Comment:
If I'm not mistaken, ESP32-C3 and -C2 do have 6 channels.
Thus if all 6 are occupied, `free_channel` will be `0b01000000`
And even with 8 channels you will end up with `0b100000000`, right?

Shouldn't this check be more like this:
```c++
int free_channel = (~ledc_handle.used_channels & (ledc_handle.used_channels+1)) & ((1<<LEDC_CHANNELS)-1));
```

Reply:
Works fine for C3 with 6 channels only. There is no need for any changes.

Reply:
I agree that the free channel calculation must take LEDC_CHANNEL into consideration.

Reply:
If bit is 1, channel is used. 0 if channel is free.

Reply:
@P-R-O-C-H-Y - PTAL.

Reply:
The LEDC_CHANNEL is taken in consideration in the structure -> ledc_handle.used_channel variable.
```
typedef struct {
    int used_channels : LEDC_CHANNELS;              // Used channels as a bits
} ledc_periph_t;
```

Reply:
Just tested that for C3 (6 channels) and ESP32 (16 channels). Both working well :)

Reply:
> The LEDC_CHANNEL is taken in consideration in the structure -> ledc_handle.used_channel variable.
> 
> ```
> typedef struct {
>     int used_channels : LEDC_CHANNELS;              // Used channels as a bits
> } ledc_periph_t;
> ```

But if you copy it to an int and invert it, when is the inversion done?
Thus is the result like this `~((int)used_channels)` or `(int)(~used_channels)` ?

Reply:
Yes, the logic works percfetly for any number of bits. Incluing empty channels in any place.
Example: mask = 0x101 ==> (NOT mask) = 0x010 ==> (mask +1) = 0x110 :: NOT & (+1) = 0x010 & 0x110 = 0x010 (channel is free).

Solved.

Reply:
My point was that there is a check for being 0 after this assignment and if you're not masking it with nr of bits, then the free_channel is not 0.

Reply:
`ledcAttachChannel` does check for `LEDC_CHANNELS`, so it will not fail indeed (apart from the possible concurrency mentioned by SuGlider)
So then it only is a matter of (possibly) never reaching the error here: `if (free_channel == 0){`

Reply:
A simple sketch for testing on C3 (6 channels):
```
#define LEDC_RESOLUTION  8
#define LEDC_FREQ     5000

void setup() {
    ledcAttach(4, LEDC_FREQ, LEDC_RESOLUTION);
    ledcAttach(5, LEDC_FREQ, LEDC_RESOLUTION);
    ledcAttach(6, LEDC_FREQ, LEDC_RESOLUTION);
    ledcAttach(7, LEDC_FREQ, LEDC_RESOLUTION);
    ledcAttach(8, LEDC_FREQ, LEDC_RESOLUTION);
    ledcAttach(9, LEDC_FREQ, LEDC_RESOLUTION);

    //No more channels available
    ledcAttach(10, LEDC_FREQ, LEDC_RESOLUTION);

    // Set the LED pin as an output
    ledcDetach(7);
    ledcAttach(10, LEDC_FREQ, LEDC_RESOLUTION);
}

void loop() {
}
```

I have added debug logs to the `ledcAttach()` function:
```
bool ledcAttach(uint8_t pin, uint32_t freq, uint8_t resolution)
{
    log_d("Used channels: %X", ledc_handle.used_channels);
    int free_channel = ~ledc_handle.used_channels & (ledc_handle.used_channels+1);
    log_d("Free channel: %u", free_channel);
    if (free_channel == 0){
        log_e("No more LEDC channels available! (maximum is %u channels)", LEDC_CHANNELS);
        return false;
    }
    uint8_t channel = __builtin_ctz(free_channel); // Convert the free_channel bit to channel number
    log_d("Channel num: %u", channel);
    return ledcAttachChannel(pin, freq, resolution, channel);
}
```

Debug output:
```
[   401][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: 0
[   407][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 1
[   412][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 0
[   418][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 4 (channel 0, resolution 8)
[   427][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: 1
[   433][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 2
[   439][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 1
[   445][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 5 (channel 1, resolution 8)
[   454][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: 3
[   460][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 4
[   465][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 2
[   471][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 6 (channel 2, resolution 8)
[   480][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: 7
[   486][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 8
[   492][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 3
[   498][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 7 (channel 3, resolution 8)
[   507][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: F
[   513][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 16
[   518][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 4
[   524][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 8 (channel 4, resolution 8)
[   533][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: 1F
[   539][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 32
[   545][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 5
[   551][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 9 (channel 5, resolution 8)
[   560][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: FFFFFFFF
[   566][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 0      <------ free channel gets to 0
[   572][E][esp32-hal-ledc.c:141] ledcAttach(): No more LEDC channels available! (maximum is 6 channels)
[   581][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: FFFFFFF7
[   588][D][esp32-hal-ledc.c:139] ledcAttach(): Free channel: 8
[   593][D][esp32-hal-ledc.c:145] ledcAttach(): Channel num: 3
[   599][I][esp32-hal-ledc.c:131] ledcAttachChannel(): LEDC attached to pin 10 (channel 3, resolution 8)
```

You can see that after using all 6 channels (0-5), the free channel is correctly calculated as 0, so the error message is printed and returned correctly.

Reply:
Now I see where I got it wrong with the bitwise algebra...
See the log line:
```
[   560][D][esp32-hal-ledc.c:137] ledcAttach(): Used channels: FFFFFFFF
```

The N-bit value is of type `int`, not `unsigned int`.
```c++
typedef struct {
    int used_channels : LEDC_CHANNELS;              // Used channels as a bits
} ledc_periph_t;
```
This thus sets it to a negative value when the "most significant" bit is set.
If you then negate it, those `F`'s become `0`'s

I've been bitten by those signed/unsigned in structs with specific bit-size quite a lot and probably will be a lot more in the future :(